evidence/reactor.go

```diff
@@ -21,233 +46,363 @@ const (
 	// Most evidence should be committed in the very next block that is why we wait
 	// just over the block production rate before sending evidence again.
 	broadcastEvidenceIntervalS = 10
-	// If a message fails wait this much before sending it again
-	peerRetryMessageIntervalMS = 100
 )
 
+type closer struct {
+	closeOnce sync.Once
+	doneCh    chan struct{}
+}
+
+func newCloser() *closer {
+	return &closer{doneCh: make(chan struct{})}
+}
+
+func (c *closer) close() {
+	c.closeOnce.Do(func() {
+		close(c.doneCh)
+	})
+}
+
 // Reactor handles evpool evidence broadcasting amongst peers.
 type Reactor struct {
-	p2p.BaseReactor
-	evpool   *Pool
-	eventBus *types.EventBus
+	service.BaseService
+
+	evpool      *Pool
+	eventBus    *types.EventBus
+	evidenceCh  *p2p.Channel
+	peerUpdates *p2p.PeerUpdatesCh
+	closeCh     chan struct{}
+
+	mtx          tmsync.RWMutex
+	peerWG       sync.WaitGroup
+	peerRoutines map[p2p.NodeID]*closer
 }
 
-// NewReactor returns a new Reactor with the given config and evpool.
-func NewReactor(evpool *Pool) *Reactor {
-	evR := &Reactor{
-		evpool: evpool,
+// NewReactor returns a reference to a new evidence reactor, which implements the
+// service.Service interface. It accepts a p2p Channel dedicated for handling
+// envelopes with EvidenceList messages.
+func NewReactor(
+	logger log.Logger,
+	evidenceCh *p2p.Channel,
+	peerUpdates *p2p.PeerUpdatesCh,
+	evpool *Pool,
+) *Reactor {
+	r := &Reactor{
+		evpool:       evpool,
+		evidenceCh:   evidenceCh,
+		peerUpdates:  peerUpdates,
+		closeCh:      make(chan struct{}),
+		peerRoutines: make(map[p2p.NodeID]*closer),
 	}
-	evR.BaseReactor = *p2p.NewBaseReactor("Evidence", evR)
-	return evR
+
+	r.BaseService = *service.NewBaseService(logger, "Evidence", r)
+	return r
 }
 
-// SetLogger sets the Logger on the reactor and the underlying Evidence.
-func (evR *Reactor) SetLogger(l log.Logger) {
-	evR.Logger = l
-	evR.evpool.SetLogger(l)
+// SetEventBus implements events.Eventable.
+func (r *Reactor) SetEventBus(b *types.EventBus) {
+	r.eventBus = b
 }
 
-// GetChannels implements Reactor.
-// It returns the list of channels for this reactor.
-func (evR *Reactor) GetChannels() []*p2p.ChannelDescriptor {
-	return []*p2p.ChannelDescriptor{
-		{
-			ID:                  EvidenceChannel,
-			Priority:            6,
-			RecvMessageCapacity: maxMsgSize,
-		},
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed. No error is returned.
+func (r *Reactor) OnStart() error {
+	go r.processEvidenceCh()
+	go r.processPeerUpdates()
+
+	return nil
+}
+
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *Reactor) OnStop() {
+	r.mtx.Lock()
+	for _, c := range r.peerRoutines {
+		c.close()
```

Comment:
Why would the router send peer updates (up/new) when reactors are shutting down?

Reply:
The router doesn't know when reactors shut down. If anything, the reactor could close the peer updates channel first, but then we'd have to make sure that peer updates channels are unbuffered and I think we'll want them to have a 1-size buffer at least (I'll explain why on the router PR, saw your comment on it).

Reply:
In any case, this was done on purpose. Consider the case:

1. `broadcastEvidenceLoop` is blocking on `r.evidenceCh.Out() <- envelope`
2. Router stops (i.e `close(r.closeCh)`) which triggers the p2p channel to close via the `defer` in `processEvidenceCh`
3. A race condition occurs between trying to send (1) while it was closed (2).

Reply:
> The router doesn't know when reactors shut down, and it has to keep running until all reactors are stopped.

How does it not know? It will have access to the routers, correct?

Reply:
> Router stops

Do you mean router or reactor? The router must run until all reactors have exited, we've set that as a basic requirement to avoid complicating reactor implementations.

Reply:
> How does it not know? It will have access to the routers, correct?

No, it only has access to the channels and peer updates subscriptions, it doesn't know anything about reactors.

Reply:
Reactor* stops.

Reply:
> I think we need to ensure all broadcastEvidenceLoop have exited before calling close(r.closeCh) and so that's why I explicitly close all of them.

But why? The whole purpose of `closeCh` is to signal to goroutines to terminate, and we explicitly check that in `broadcastEvidenceLoop` -- why would we need to shut down goroutines before signalling them to shut down? 🙂

Reply:
Because of the race condition between `defer r.evidenceCh.Close()` and `r.evidenceCh.Out() <- envelope`.

Reply:
Ah crap, you're right. Should've read your comments more carefully, sorry. We'll need to close the peer updates subscription first then, I think.

I don't think the select is necessary anyway though, since we'll have to wait for the waitgroup before closing.

Reply:
> We'll need to close the peer updates subscription first then, I think.

We can't do this either because we use a single `closeCh` to signal to _all_ goroutines to exit (i.e. we don't have a unique `closeCh` per goroutine/domain). Although we could, I think it's not necessary and overkill.

Instead, I've opted for a simple `r.IsRunning()` check in `processPeerUpdate`.

Reply:
That'll work 👍